### PR TITLE
Replace moment with date-fns in hca

### PR DESCRIPTION
**Pull request by bkjohnson**
## Description

This is part of https://github.com/department-of-veterans-affairs/va.gov-team/issues/6763

This removes all _explicit_ usage of `moment` from the `hca` app.  The forms system still uses moment, so it will still get used implicitly in the app.

## Testing done

Local unit testing

## Acceptance criteria
- [ ]

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: src/applications/hca/enrollment-status-helpers.jsx [hunk 1]
```diff
@@ -89,7 +89,7 @@ export function getEnrollmentDetails(
     blocks.push(
       <>
         <strong>You applied on: </strong>
-        {moment(applicationDate).format('MMMM D, YYYY')}
+        {format(Date.parse(applicationDate), 'MMMM d, yyyy')}
```
Comment (bkjohnson) [inline src/applications/hca/enrollment-status-helpers.jsx:92]: `applicationDate` is the first argument in `getEnrollmentDetails`. Even though there isn't a unit test specifically for this function, it is used in `getWarningStatus()`, and the second argument is passed directly through.

https://github.com/department-of-veterans-affairs/vets-website/blob/8af6da8f01587d87d66e12ddde79b0e4f22e81b7/src/applications/hca/tests/enrollment-status-helpers.unit.spec.js#L40-L45

File: src/applications/hca/enrollment-status-helpers.jsx [hunk 2]
```diff
@@ -98,7 +98,7 @@ export function getEnrollmentDetails(
     blocks.push(
       <>
         <strong>We enrolled you on: </strong>
-        {moment(enrollmentDate).format('MMMM D, YYYY')}
+        {format(Date.parse(enrollmentDate), 'MMMM d, yyyy')}
```
Comment (bkjohnson) [inline src/applications/hca/enrollment-status-helpers.jsx:101]: `enrollmentDate` has the same story as `applicationDate`.  It may be wise to add some unit tests for this function.

File: src/applications/hca/enrollment-status-helpers.jsx [hunk 3]
```diff
@@ -971,7 +971,7 @@ export function getAlertContent(
     blocks.push(
       <p key="you-applied-on">
         <strong>You applied on:</strong>{' '}
-        {moment(applicationDate).format('MMMM D, YYYY')}
+        {format(Date.parse(applicationDate), 'MMMM d, yyyy')}
```
Comment (bkjohnson) [inline src/applications/hca/enrollment-status-helpers.jsx:974]: This is in an untested function.  The only place it is used is in `src/applications/personalization/dashboard/components/HCAStatusAlert.jsx`

File: src/applications/hca/tests/containers/ConfirmationPage.unit.spec.jsx [hunk 4]
```diff
@@ -26,7 +26,7 @@ describe('hca <ConfirmationPage>', () => {
     const tree = SkinDeep.shallowRender(<ConfirmationPage form={form} />);
 
     expect(tree.subTree('.claim-list')).to.exist;
-    expect(tree.everySubTree('span')[2].text()).to.contain('Jan. 1, 2010');
+    expect(tree.everySubTree('span')[2].text()).to.contain('Jan 1, 2010');
```
Comment (bkjohnson) [inline src/applications/hca/tests/containers/ConfirmationPage.unit.spec.jsx:29]: The formatting for abbreviated months in `date-fns` does not include a period for truncated months.

File: src/applications/hca/validation.js [hunk 5]
```diff
@@ -31,19 +28,20 @@ export function validateServiceDates(
   // TODO: Use a constant instead of a magic string
   if (
     !isValidDateRange(fromDate, toDate) ||
-    moment(lastDischargeDate, 'YYYY-MM-DD').isAfter(endDateInfo.endDate)
+    isAfter(Date.parse(lastDischargeDate), endDateInfo.endDate)
```
Comment (bkjohnson) [inline src/applications/hca/validation.js:31]: `lastDischargeDate` is an ISO date string.

https://github.com/department-of-veterans-affairs/vets-website/blob/8f57a94c2e86cc81b6d388d0af41c64af6286184/src/applications/hca/tests/validation.unit.spec.js#L20-L29

File: src/applications/hca/validation.js [hunk 6]
```diff
@@ -31,19 +28,20 @@ export function validateServiceDates(
...
   ) {
     errors.lastDischargeDate.addError(
-      `Discharge date must be after the service period start date and before ${endDateInfo.endDate.format(
-        'MMMM D, YYYY',
+      `Discharge date must be after the service period start date and before ${format(
+        endDateInfo.endDate,
```
Comment (bkjohnson) [inline src/applications/hca/validation.js:35]: `endDateInfo.endDate` is now a native Date object.

File: src/applications/hca/validation.js [hunk 7]
```diff
@@ -31,19 +28,20 @@ export function validateServiceDates(
...
+        'MMMM d, yyyy',
       )} (${endDateInfo.description} from today)`,
     );
   }
 
   if (veteranDateOfBirth) {
-    const dateOfBirth = moment(veteranDateOfBirth);
+    const dateOfBirth = Date.parse(veteranDateOfBirth);
```
Comment (bkjohnson) [inline src/applications/hca/validation.js:42]: `veteranDateOfBirth` is also an ISO date string.

File: src/applications/hca/validation.js [hunk 8]
```diff
@@ -31,19 +28,20 @@ export function validateServiceDates(
...
 
-    if (dateOfBirth.add(15, 'years').isAfter(moment(lastEntryDate))) {
+    if (isAfter(addYears(dateOfBirth, 15), Date.parse(lastEntryDate))) {
```
Comment (bkjohnson) [inline src/applications/hca/validation.js:44]: `lastEntryDate` is also an ISO date string.

File: src/applications/hca/validation.js [hunk 9]
```diff
@@ -56,13 +54,13 @@ export function validateMarriageDate(
   marriageDate,
   { spouseDateOfBirth, veteranDateOfBirth, discloseFinancialInformation },
 ) {
-  const vetDOB = moment(veteranDateOfBirth);
-  const spouseDOB = moment(spouseDateOfBirth);
-  const marriage = moment(marriageDate);
+  const vetDOB = Date.parse(veteranDateOfBirth);
+  const spouseDOB = Date.parse(spouseDateOfBirth);
+  const marriage = Date.parse(marriageDate);
```
Comment (bkjohnson) [inline src/applications/hca/validation.js:59]: `marriageDate` is the second argument in `validateMarriageDate`.  All of these three arguments are ISO date strings.

https://github.com/department-of-veterans-affairs/vets-website/blob/8f57a94c2e86cc81b6d388d0af41c64af6286184/src/applications/hca/tests/validation.unit.spec.js#L93-L97

File: src/applications/hca/validation.js [hunk 10]
```diff
@@ -79,10 +77,11 @@ export function validateDependentDate(
   messages,
   index,
 ) {
-  const dependent = moment(dependentDate);
-  const dob = moment(_.get(`dependents[${index}].dateOfBirth`, formData));
+  const dependent = Date.parse(dependentDate);
```
Comment (bkjohnson) [inline src/applications/hca/validation.js:80]: `dependentDate` is the second argument in `validateDependentDate`, and it is also an ISO date string.

https://github.com/department-of-veterans-affairs/vets-website/blob/8f57a94c2e86cc81b6d388d0af41c64af6286184/src/applications/hca/tests/validation.unit.spec.js#L131-L145

File: src/applications/hca/validation.js [hunk 11]
```diff
@@ -79,10 +77,11 @@ export function validateDependentDate(
...
+  const dob =
+    Date.parse(formData?.dependents?.[index]?.dateOfBirth) || Date.now();
```
Comment (bkjohnson) [inline src/applications/hca/validation.js:82]: This is a bit confusing. The arguments in the previous `_.get()` call seem swapped - [the object should be first and the path second](https://lodash.com/docs/4.17.15#get).

However, the `_.get()` call was _likely_ resulting in `undefined` each time, and [`moment(undefined)` returns the current datetime](https://momentjs.com/docs/#/parsing/now/):

> Note: Function parameters default to undefined when not passed in. Moment treats moment(undefined) as moment().

**Review by bkjohnson — COMMENTED** [review]
Adding some context.

File: src/applications/hca/tests/containers/ConfirmationPage.unit.spec.jsx (hunk 4 above)
Comment (rianfowler) [reply src/applications/hca/tests/containers/ConfirmationPage.unit.spec.jsx:29]: The [content style guide](https://design.va.gov/content-style-guide/dates-and-numbers#dates-and-years) indicates they have to have a period. I [think](https://date-fns.org/v2.11.1/docs/format#examples) we can use the escape character to add a period in the format string. Maybe 

```
{format(parseISO(response.timestamp), "MMM'.' d, yyyy")}
```

**Review by rianfowler — CHANGES_REQUESTED** [review]

File: src/applications/hca/tests/containers/ConfirmationPage.unit.spec.jsx (hunk 4 above)
Comment (bkjohnson) [reply src/applications/hca/tests/containers/ConfirmationPage.unit.spec.jsx:29]: I'll mess around with this - we may have some edge cases to account for with shorter months (i.e. `March. 14, 2020`), so I'll see how it handles those.

File: src/applications/hca/validation.js (hunk 11 above)
Comment (bkjohnson) [reply src/applications/hca/validation.js:82]: if the optional chaining returns `undefined`, `Date.parse(undefined)` will be `NaN`, so the `|| Date.now()` should achieve the same thing that was previously happening.

File: src/applications/hca/tests/containers/ConfirmationPage.unit.spec.jsx (hunk 4 above)
Comment (rianfowler) [reply src/applications/hca/tests/containers/ConfirmationPage.unit.spec.jsx:29]: That true - we aren't supposed to abbreviate those. 
external-content duckduckgo-9

File: src/applications/hca/validation.js (hunk 11 above)
Comment (rianfowler) [reply src/applications/hca/validation.js:82]: >The arguments in the previous _.get() call seem swapped

It's [lodash/fp](https://gist.github.com/jfmengels/6b973b69c491375117dc#_getpath-object).

> if the optional chaining returns undefined, Date.parse(undefined) will be NaN, so the || Date.now() should achieve the same thing that was previously happening

This looks fine. maybe add a comment explaining that `Date.parse` returns `NaN` when applied to `undefined`.

File: src/applications/hca/tests/containers/ConfirmationPage.unit.spec.jsx (hunk 4 above)
Comment (bkjohnson) [reply src/applications/hca/tests/containers/ConfirmationPage.unit.spec.jsx:29]: I've seen some code that Eugene wrote to handle this exact thing for downtime notifications.

Comment (bkjohnson) [reply src/applications/hca/tests/containers/ConfirmationPage.unit.spec.jsx:29]: [Closing for now](https://github.com/department-of-veterans-affairs/va.gov-team/issues/6763#issuecomment-609983881).  The extra work to ensure that _every_ format call meets the content guidelines isn't worth it in my opinion. I'd like for `date-fns` to provide a way of changing the default locale so we have an easier way of controlling things like this.

**Comment by bkjohnson** [comment]
Closing for now: https://github.com/department-of-veterans-affairs/vets-website/pull/12112#discussion_r404327609
